psydac/core/bsplines.py

```diff
@@ -482,7 +482,9 @@ def breakpoints( knots, degree ):
         Abscissas of all breakpoints.
 
     """
-    return np.unique( knots[degree:-degree] )
+    knots = np.array(knots)
+    diff  = np.append(True, np.diff(knots[degree:-degree])>1e-15)
+    return knots[degree:-degree][diff]
```

Comment:
I do not agree with this approach of hiding comparison tolerances inside the code. How can one know if `1e-15` is sufficiently small? If you want to do this you should provide the optional argument `tol` for the user to change.

In any case I don't think that the `breakpoints` function should take care of this. The original implementation was assuming that a knot sequence containing repeated knots is usually created by copying the same knot value multiple times, so that they will be identical no matter the precision. If you are worried about this you can add the following two things:

1. A dedicated function that "sanitizes" the knot sequence by applying a user-provided tolerance;
2. A clear comment in the docstring of `breakpoints`, which states that the input knot sequence must be properly sanitized.

Another approach, which I think is bullet-proof, is to use knot sequences without repeated knots in conjunction with an integer array containing the multiplicities of each knot.

Reply:
when we have repeated knots, we have the choice to use `numpy.unique` to get the break points, or the way that I did it,I'm not sure if it's good to use numpy.unique with float numbers , that's why I used this method with a small enough tolerance.
so I suggest to add a keywrod argument that specifies the tolerance